crypto-ffi/src/wasm/mod.rs

```diff
@@ -79,96 +79,201 @@ pub extern "C" fn __stack_chk_fail() {
-                WasmError::EnumError => "EnumError",
-                WasmError::TransactionFailed { .. } => "TransactionFailed",
+// This implementation is intended to be temporary; we're going to be completely restructuring the way we handle
+// errors in `core-crypto` soon. We can replace this with better error patterns when we do.
+//
+// Certain error mappings could apply to both MLS and Proteus. In all such cases, we map them to the MLS variant.
+// When we redesign the errors in `core-crypto`, these ambiguities should disappear anyway.
+impl From<CryptoError> for InternalError {
+    fn from(value: CryptoError) -> Self {
+        #[cfg(feature = "proteus")]
+        if let Some(error_code) = value.proteus_error_code() {
+            return match error_code {
+                102 => ProteusError::SessionNotFound, // reviewers take note: is this correct?
+                204 => ProteusError::RemoteIdentityChanged,
+                209 => ProteusError::DuplicateMessage,
+                _ => ProteusError::Other(Some(error_code)),
             }
-            .to_string(),
-            message: e.0.to_string(),
-            rust_stack_trace: format!("{:?}", e.0),
-            proteus_error_code: match e.0 {
-                WasmError::TransactionFailed { proteus_error_code, .. } => proteus_error_code,
-                _ => e.proteus_error_code(),
-            },
+            .into();
+        }
+
+        match value {
+            CryptoError::ConversationAlreadyExists(id) => MlsError::ConversationAlreadyExists(id).into(),
+            CryptoError::DuplicateMessage => MlsError::DuplicateMessage.into(),
+            CryptoError::BufferedFutureMessage => MlsError::BufferedFutureMessage.into(),
+            CryptoError::WrongEpoch => MlsError::WrongEpoch.into(),
+            CryptoError::MessageEpochTooOld => MlsError::MessageEpochTooOld.into(),
+            CryptoError::SelfCommitIgnored => MlsError::SelfCommitIgnored.into(),
+            CryptoError::UnmergedPendingGroup => MlsError::UnmergedPendingGroup.into(),
+            CryptoError::StaleProposal => MlsError::StaleProposal.into(),
+            CryptoError::StaleCommit => MlsError::StaleCommit.into(),
+            CryptoError::E2eiError(_) => Self::E2eiError,
+            _ => MlsError::Other.into(),
         }
     }
 }
 
-#[derive(Debug, thiserror::Error)]
-pub struct CoreCryptoError(#[from] WasmError);
+impl From<E2eIdentityError> for InternalError {
+    fn from(_: E2eIdentityError) -> Self {
+        Self::E2eiError
+    }
+}
 
-impl CoreCryptoError {
-    fn proteus_error_code(&self) -> u32 {
-        let WasmError::CryptoError(e) = &self.0 else {
-            return 0;
-        };
+impl InternalError {
+    fn variant_name(&self) -> String {
+        let mut out = self.as_ref().to_string();
+        match self {
+            Self::MlsError(mls) => out += mls.as_ref(),
+            Self::ProteusError(proteus) => out += proteus.as_ref(),
```

Comment:
How does the resulting `out` string look like? This code just concatenates two strings, without any separator.
Can you provide a few examples?